src/olympia/blocklist/tests/test_mlbf.py

```diff
@@ -300,44 +365,105 @@ def test_diff_block_added_and_removed(self):
+                MLBF.hash_filter_inputs(
+                    [(block_version.block.guid, block_version.version.version)]
+                ),
+                1,
+            ),
+            BlockType.SOFT_BLOCKED: (
+                MLBF.hash_filter_inputs(
+                    [(block_version.block.guid, block_version.version.version)]
+                ),
+                [],
+                1,
+            ),
+        }
+
+    def test_diff_block_soft_to_hard(self):
+        addon, block = self._blocked_addon()
+        block_version = self._block_version(
+            block, self._version(addon), block_type=BlockType.SOFT_BLOCKED
+        )
+        base_mlbf = MLBF.generate_from_db('test')
+        block_version.update(block_type=BlockType.BLOCKED)
+        next_mlbf = MLBF.generate_from_db('test_two')
+
+        assert next_mlbf.generate_diffs(previous_mlbf=base_mlbf) == {
+            BlockType.BLOCKED: (
+                MLBF.hash_filter_inputs(
+                    [(block_version.block.guid, block_version.version.version)]
+                ),
+                [],
+                1,
+            ),
+            BlockType.SOFT_BLOCKED: (
+                [],
+                MLBF.hash_filter_inputs(
+                    [(block_version.block.guid, block_version.version.version)]
+                ),
+                1,
+            ),
+        }
+
+    def test_generate_stash_returns_expected_stash(self):
+        addon, block = self._blocked_addon()
+        block_versions = [
+            self._block_version(block, self._version(addon)) for _ in range(10)
+        ]
         mlbf = MLBF.generate_from_db('test')
         mlbf.generate_and_write_stash()
 
         with mlbf.storage.open(mlbf.stash_path, 'r') as f:
             assert json.load(f) == {
                 'blocked': MLBF.hash_filter_inputs(
-                    [(block_version.block.guid, block_version.version.version)]
+                    [
+                        (block_version.block.guid, block_version.version.version)
+                        for block_version in block_versions
+                    ]
                 ),
                 'unblocked': [],
             }
-        block_version.delete()
+
+        for i, block_version in enumerate(block_versions):
+            if i % 2 == 0:
+                block_version.delete()
 
         next_mlbf = MLBF.generate_from_db('test_two')
         next_mlbf.generate_and_write_stash(previous_mlbf=mlbf)
         with next_mlbf.storage.open(next_mlbf.stash_path, 'r') as f:
             assert json.load(f) == {
                 'blocked': [],
                 'unblocked': MLBF.hash_filter_inputs(
-                    [(block_version.block.guid, block_version.version.version)]
+                    [
+                        (block_version.block.guid, block_version.version.version)
+                        for i, block_version in enumerate(block_versions)
+                        if i % 2 == 0
+                    ]
```

Comment:
This is very hard to follow what's going on and what is being asserted. I'd suggest you delete one or two `block_version` before generating the next mlbf, and then you use that instead, but without a loop or modulo.